### ISSUE TITLE
acl_manager timing out every minute

### ISSUE THREAD

**Issue by liljenstolpe** [Metaswitch/calico-docker-prototype#18]
This is with only core-01, 

log from acl_manager:
```
2015-01-19 19:40:22,874 [INFO] calico.acl_manager.acl_manager 76: pyzmq version is 14.0.1
2015-01-19 19:40:22,876 [INFO] calico.acl_manager.acl_publisher 126: Sending ACL heartbeat {"issued": 1421696422876.61, "type": "HEARTBEAT"}
2015-01-19 19:40:22,878 [INFO] calico.acl_manager.net_subscriber 91: Received start of day response {"message": "Hooray", "type": "GETGROUPS", "rc": "SUCCESS"}
2015-01-19 19:40:27,015 [INFO] calico.acl_manager.acl_publisher 95: ACL Manager received packet {u'type': u'HEARTBEAT'} from ^@k<8B>Eg
2015-01-19 19:40:27,015 [WARNING] calico.acl_manager.acl_publisher 105: Received query {u'type': u'HEARTBEAT'} of unknown type
2015-01-19 19:40:32,039 [INFO] calico.acl_manager.acl_publisher 95: ACL Manager received packet {u'type': u'HEARTBEAT'} from ^@k<8B>Eg
2015-01-19 19:40:32,039 [WARNING] calico.acl_manager.acl_publisher 105: Received query {u'type': u'HEARTBEAT'} of unknown type
2015-01-19 19:40:37,048 [INFO] calico.acl_manager.acl_publisher 95: ACL Manager received packet {u'type': u'HEARTBEAT'} from ^@k<8B>Eg
2015-01-19 19:40:37,048 [WARNING] calico.acl_manager.acl_publisher 105: Received query {u'type': u'HEARTBEAT'} of unknown type
2015-01-19 19:40:42,066 [INFO] calico.acl_manager.acl_publisher 95: ACL Manager received packet {u'type': u'HEARTBEAT'} from ^@k<8B>Eg
2015-01-19 19:40:42,066 [WARNING] calico.acl_manager.acl_publisher 105: Received query {u'type': u'HEARTBEAT'} of unknown type
2015-01-19 19:40:47,075 [INFO] calico.acl_manager.acl_publisher 95: ACL Manager received packet {u'type': u'HEARTBEAT'} from ^@k<8B>Eg
2015-01-19 19:40:47,075 [WARNING] calico.acl_manager.acl_publisher 105: Received query {u'type': u'HEARTBEAT'} of unknown type
2015-01-19 19:40:52,083 [INFO] calico.acl_manager.acl_publisher 95: ACL Manager received packet {u'type': u'HEARTBEAT'} from ^@k<8B>Eg
2015-01-19 19:40:52,083 [WARNING] calico.acl_manager.acl_publisher 105: Received query {u'type': u'HEARTBEAT'} of unknown type
2015-01-19 19:40:52,886 [INFO] calico.acl_manager.acl_publisher 126: Sending ACL heartbeat {"issued": 1421696452886.102, "type": "HEARTBEAT"}
2015-01-19 19:40:57,403 [INFO] calico.acl_manager.acl_publisher 95: ACL Manager received packet {u'type': u'HEARTBEAT'} from ^@k<8B>Eg
2015-01-19 19:40:57,403 [WARNING] calico.acl_manager.acl_publisher 105: Received query {u'type': u'HEARTBEAT'} of unknown type
2015-01-19 19:41:02,582 [INFO] calico.acl_manager.acl_publisher 95: ACL Manager received packet {u'type': u'HEARTBEAT'} from ^@k<8B>Eg
2015-01-19 19:41:02,582 [WARNING] calico.acl_manager.acl_publisher 105: Received query {u'type': u'HEARTBEAT'} of unknown type
2015-01-19 19:41:07,591 [INFO] calico.acl_manager.acl_publisher 95: ACL Manager received packet {u'type': u'HEARTBEAT'} from ^@k<8B>Eg
2015-01-19 19:41:07,591 [WARNING] calico.acl_manager.acl_publisher 105: Received query {u'type': u'HEARTBEAT'} of unknown type
2015-01-19 19:41:12,601 [INFO] calico.acl_manager.acl_publisher 95: ACL Manager received packet {u'type': u'HEARTBEAT'} from ^@k<8B>Eg
2015-01-19 19:41:12,601 [WARNING] calico.acl_manager.acl_publisher 105: Received query {u'type': u'HEARTBEAT'} of unknown type
2015-01-19 19:41:17,613 [INFO] calico.acl_manager.acl_publisher 95: ACL Manager received packet {u'type': u'HEARTBEAT'} from ^@k<8B>Eg
2015-01-19 19:41:17,614 [WARNING] calico.acl_manager.acl_publisher 105: Received query {u'type': u'HEARTBEAT'} of unknown type
2015-01-19 19:41:22,626 [INFO] calico.acl_manager.acl_publisher 95: ACL Manager received packet {u'type': u'HEARTBEAT'} from ^@k<8B>Eg
2015-01-19 19:41:22,626 [WARNING] calico.acl_manager.acl_publisher 105: Received query {u'type': u'HEARTBEAT'} of unknown type
2015-01-19 19:41:22,910 [INFO] calico.acl_manager.acl_publisher 126: Sending ACL heartbeat {"issued": 1421696482910.792, "type": "HEARTBEAT"}
2015-01-19 19:41:27,635 [INFO] calico.acl_manager.acl_publisher 95: ACL Manager received packet {u'type': u'HEARTBEAT'} from ^@k<8B>Eg
2015-01-19 19:41:27,635 [WARNING] calico.acl_manager.acl_publisher 105: Received query {u'type': u'HEARTBEAT'} of unknown type
2015-01-19 19:41:27,940 [ERROR] calico.acl_manager.net_subscriber 100: ACL Manager plugin heartbeat timed out - exiting
2015-01-19 19:41:28,254 [INFO] calico.acl_manager.acl_manager 76: pyzmq version is 14.0.1
2015-01-19 19:41:28,256 [INFO] calico.acl_manager.acl_publisher 126: Sending ACL heartbeat {"issued": 1421696488255.9878, "type": "HEARTBEAT"}
2015-01-19 19:41:28,257 [INFO] calico.acl_manager.net_subscriber 91: Received start of day response {"message": "Hooray", "type": "GETGROUPS", "rc": "SUCCESS"}
2015-01-19 19:41:32,644 [INFO] calico.acl_manager.acl_publisher 95: ACL Manager received packet {u'type': u'HEARTBEAT'} from ^@k<8B>Eg
```

docker ps output
```
core-01 demo # docker ps
CONTAINER ID        IMAGE               COMMAND                CREATED             STATUS              PORTS               NAMES
1d57f3421467        calico:felix        "calico-acl-manager    4 minutes ago       Up 35 seconds                           aclmgr
cfc1663f20eb        calico:felix        "calico-felix --conf   4 minutes ago       Up 4 minutes                            felix
ad5d786a2168        calico:plugin       "python /opt/scripts   5 minutes ago       Up 5 minutes                            plugin2
cad19d7c9e04        calico:plugin       "python /opt/scripts   5 minutes ago       Up 5 minutes                            plugin1
dc8c597c6290        calico:bird         "/usr/bin/run_bird b   5 minutes ago       Up 5 minutes                            bird
core-01 demo # less /var/log/calico/acl_manager.log
core-01 demo # docker ps
CONTAINER ID        IMAGE               COMMAND                CREATED             STATUS              PORTS               NAMES
1d57f3421467        calico:felix        "calico-acl-manager    7 minutes ago       Up 19 seconds                           aclmgr
cfc1663f20eb        calico:felix        "calico-felix --conf   7 minutes ago       Up 7 minutes                            felix
ad5d786a2168        calico:plugin       "python /opt/scripts   8 minutes ago       Up 8 minutes                            plugin2
cad19d7c9e04        calico:plugin       "python /opt/scripts   8 minutes ago       Up 8 minutes                            plugin1
dc8c597c6290        calico:bird         "/usr/bin/run_bird b   8 minutes ago       Up 8 minutes                            bird
```

**Comment by Lukasa** [comment]
To be clear, what's happening here is that the ACL manager is waiting on a heartbeat from the plugin that it never receives. Here's how this is supposed to look (from a functioning OpenStack system's logs):

```
2015-01-20 09:37:19,358 [INFO] calico.acl_manager.acl_publisher 95: ACL Manager received packet {u'type': u'HEARTBEAT'} from ^@k<8b>Eg
2015-01-20 09:37:19,359 [WARNING] calico.acl_manager.acl_publisher 105: Received query {u'type': u'HEARTBEAT'} of unknown type
2015-01-20 09:37:19,445 [INFO] calico.acl_manager.acl_publisher 126: Sending ACL heartbeat {"issued": 1421746639445.889, "type": "HEARTBEAT"}
2015-01-20 09:37:20,949 [INFO] calico.acl_manager.acl_publisher 95: ACL Manager received packet {u'type': u'HEARTBEAT'} from ^@k<8b>Eh
2015-01-20 09:37:20,949 [WARNING] calico.acl_manager.acl_publisher 105: Received query {u'type': u'HEARTBEAT'} of unknown type
2015-01-20 09:37:21,450 [INFO] calico.acl_manager.acl_publisher 95: ACL Manager received packet {u'type': u'HEARTBEAT'} from ^@k<8b>Eg
2015-01-20 09:37:21,450 [WARNING] calico.acl_manager.acl_publisher 105: Received query {u'type': u'HEARTBEAT'} of unknown type
2015-01-20 09:37:22,027 [INFO] calico.acl_manager.net_subscriber 106: Received published message {u'issued': 1421746642026.855, u'type': u'HEARTBEAT'}
2015-01-20 09:37:22,027 [INFO] calico.acl_manager.net_subscriber 109: Network API Heartbeat received
```

**Comment by Lukasa** [comment]
This is therefore not an ACL manager bug, it's a plugin bug.

**Comment by ed-harrison** [comment]
I would claim the ACL Manager logging is bugged, though.  I would suggest that

-  each log should be clearer which heartbeats are being logged (i.e. ones from Felix or the Plug-in)
-  it seems wrong for a log to state that a Heartbeat is of "unknown type" (surely it is a known type - Heartbeat)

**Comment by plwhite** [comment]
Regarding Ed's comment

* ACL Manager is trying hard to log this - the log is broken as noted in https://github.com/Metaswitch/calico/issues/122.
* The failure to correctly handle the heartbeat is https://github.com/Metaswitch/calico/issues/122 (though there is some, in my view crazy, debate about the right solution).

**Comment by plwhite** [comment]
It'll be hard to investigate further without logs (we need the diagnostics logs as collected by the diags tool; just having one side of the exchange doesn't tell us why the far end isn't responding).

The most likely cause is one of : 

* The plugin is down (unlikely given the `docker ps` output)
* The IP address of the controller as specified in the ACL Manager configuration is wrong (unlikely given that a GETGROUPS message happened earlier).

That leaves the possibility that the plugin is doing something wrong and failing to send heartbeats, but there are no plugin logs to answer that question.

**Comment by plwhite** [comment]
Aha. I think this is because the *network* plugin does not send heartbeats, and so if there is no configured data, then the ACL Manager recycles.

This is benign, but well worth fixing to avoid log confusion, so I'll fix.

**Comment by plwhite** [comment]
Fixed in PR #26.
